lib/Attachment.php

```diff
@@ -0,0 +1,556 @@
+	 *     <span class="download-info">(Download, {{ attachment.size }})</span>
+	 * </a>
+	 * ```
+	 * @return mixed|null The filesize string in a human readable format.
+	 */
+	public function size() {
+		if ( ! $this->file_size ) {
+			$formatted_size  = size_format( $this->size_raw() );
+			$this->file_size = str_replace( ' ', '&nbsp;', $formatted_size );
+		}
+
+		return $this->file_size;
+	}
+
+	/**
+	 * Gets filesize in bytes.
+	 *
+	 * @api
+	 * @since 2.0.0
+	 * @example
+	 *
+	 * ```twig
+	 * <table>
+	 *     {% for attachment in Attachment(attachment_ids) %}
+	 *         <tr>
+	 *             <td>{{ attachment.title }}</td>
+	 *             <td>{{ attachment.extension }}</td>
+	 *             <td>{{ attachment.size_raw }} bytes</td>
+	 *         </tr>
+	 *     {% endfor %}
+	 * </table>
+	 * ```
+	 *
+	 * @return mixed|null The filesize string in bytes, or false if the filesize can’t be read.
+	 */
+	public function size_raw() {
+		if ( ! $this->file_size_raw ) {
+			$this->file_size_raw = filesize( $this->file_loc );
+		}
+
+		return $this->file_size_raw;
+	}
+
+	/**
+	 * Gets the extension of the attached file.
+	 *
+	 * @api
+	 * @since 2.0.0
+	 * @example
+	 *
+	 * Use extension information in a link that downloads a file:
+	 *
+	 * ```twig
+	 * <a class="download" href="{{ attachment.src }}" download="{{ attachment.title }}">
+	 *     <span class="download-title">{{ attachment.title }}</span>
+	 *     <span class="download-info">
+	 *         (Download {{ attachment.extension|upper }}, {{ attachment.size }})
+	 *     </span>
+	 * </a>
+	 * ```
+	 *
+	 * @return null|string An uppercase extension string.
+	 */
+	public function extension() {
+		if ( ! $this->file_extension ) {
+			$file_info = wp_check_filetype( $this->file );
+
+			if ( ! empty( $file_info['ext'] ) ) {
+				$this->file_extension = strtoupper( $file_info['ext'] );
+			}
+		}
+
+		return $this->file_extension;
+	}
+
+	/**
+	 * Gets the parent object.
+	 *
+	 * The parent object of an attachment is a post it is assigned to.
+	 *
+	 * @api
+	 * @example
+	 * ```twig
+	 * This image is assigned to {{ image.parent.title }}
+	 * ```
+	 *
+	 * @return false|\Timber\Post Parent object as a `Timber\Post`. Returns `false` if no parent
+	 *                            object is defined.
+	 */
+	public function parent() {
+		if ( ! $this->post_parent ) {
+			return false;
+		}
+
+		return new $this->PostClass( $this->post_parent );
+	}
+
+	/**
+	 * Get a PHP array with pathinfo() info from the file
+	 *
+	 * @api
+	 *
+	 * @return array
+	 */
+	public function get_pathinfo() {
+		return pathinfo( $this->file );
+	}
+
+	/**
+	 * Returns the `wp_upload_dir` and saves result to static var
+	 *
+	 * @internal
+	 * @link https://developer.wordpress.org/reference/functions/wp_upload_dir/
+	 *
+	 * @return array
+	 */
+	public static function wp_upload_dir() {
+		static $wp_upload_dir = false;
+
+		if ( ! $wp_upload_dir ) {
+			$wp_upload_dir = wp_upload_dir();
```

Comment:
Maybe refactor to `wp_get_upload_dir`. See https://github.com/WordPress/WordPress/blob/master/wp-includes/functions.php#L2003 for more information

Reply:
Also no static cache like implemented would be needed as WP does this by itself. See https://github.com/WordPress/WordPress/blob/master/wp-includes/functions.php#L2057